pennylane_lightning/core/src/measurements/MeasurementsBase.hpp

```diff
@@ -111,6 +114,183 @@ template <class StateVectorT, class Derived> class MeasurementsBase {
     auto generate_samples(size_t num_samples) -> std::vector<size_t> {
         return static_cast<Derived *>(this)->generate_samples(num_samples);
     };
-};
 
-} // namespace Pennylane::Measures
\ No newline at end of file
+    /**
+     * @brief Calculate the expectation value for a general Observable.
+     *
+     * @param obs Observable.
+     * @param num_shots Number of shots used to generate samples
+     * @param shot_range The range of samples to use. All samples are used
+     * by default.
+     *
+     * @return Expectation value with respect to the given observable.
+     */
+    auto expval(const Observable<StateVectorT> &obs, const size_t &num_shots,
+                const std::vector<size_t> &shot_range = {}) -> PrecisionT {
+        PrecisionT result = 0;
+
+        if (obs.getObsName().find("SparseHamiltonian") != std::string::npos) {
+            PL_ABORT("For SparseHamiltonian Observables, expval calculation is "
+                     "not supported by shots");
+        } else if (obs.getObsName().find("Hermitian") != std::string::npos) {
+            PL_ABORT("For Hermitian Observables, expval calculation is not "
+                     "supported by shots");
+        } else if (obs.getObsName().find("Hamiltonian") != std::string::npos) {
+            auto coeffs = obs.getCoeffs();
+            for (size_t obs_term_idx = 0; obs_term_idx < coeffs.size();
+                 obs_term_idx++) {
+                auto obs_samples = measure_with_samples(
+                    obs, num_shots, shot_range, obs_term_idx);
+                PrecisionT result_per_term = std::accumulate(
+                    obs_samples.begin(), obs_samples.end(), 0.0);
+
+                result +=
+                    coeffs[obs_term_idx] * result_per_term / obs_samples.size();
+            }
+        } else {
+            auto obs_samples = measure_with_samples(obs, num_shots, shot_range);
+            result =
+                std::accumulate(obs_samples.begin(), obs_samples.end(), 0.0);
+            result = result / obs_samples.size();
+        }
+        return result;
+    }
+
+    /**
+     * @brief Calculate the expectation value for a general Observable.
+     *
+     * @param obs Observable.
+     * @param num_shots Number of shots used to generate samples
+     * @param shot_range The range of samples to use. All samples are used
+     * by default.
+     * @param term_idx Index of a Hamiltonian term
+     *
+     * @return Expectation value with respect to the given observable.
+     */
+    auto measure_with_samples(const Observable<StateVectorT> &obs,
+                              const size_t &num_shots,
+                              const std::vector<size_t> &shot_range,
+                              size_t term_idx = 0) {
+        const size_t num_qubits = _statevector.getTotalNumQubits();
+        std::vector<size_t> obs_wires;
+        std::vector<size_t> identity_wires;
```

Comment:
Can we set the capacity of these vectors before passing by ref to `_sample_state`?

Reply:
Unfortunately, we can't. The capacity of `identity_wires` is unknown and can be only set when an obs is parsed. While the main reason we can't set the capacity of obs_wires is that `obs_wires` of Hamiltonian term is not accessible.